### ISSUE TITLE
[布林带均值回归策略] | Bollinger Band Reversion

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4436]
**标签**: quantitative-strategy, technical-analysis, bollinger-bands, mean-reversion
### 摘要
使用布林带识别价格偏离，结合RSI进行均值回归交易，捕捉价格对均值的周期性回归特性。
### 算法描述
布林带（Bollinger Bands）是经典的技术分析工具，由中轨（MA）和上下轨（MA±2σ）组成。价格触及布林带边缘后均值回归的概率较高。

**核心原理**：根据统计学原理，价格序列服从正态分布时，约95%的价格应落在均值±2倍标准差区间内。当价格触及布林带边缘时，出现显著偏离，存在均值回归动力。

**数学定义**：

布林带：
$$\text{Middle}_t = \text{MA}(P_t, n)$$
$$\text{Upper}_t = \text{Middle}_t + k \cdot \sigma_t$$
$$\text{Lower}_t = \text{Middle}_t - k \cdot \sigma_t$$

其中 $n$ 为移动平均周期（通常20），$k$ 为标准差倍数（通常2），$\sigma_t$ 为滚动标准差。

RSI指标：
$$\text{RSI}_t = 100 - \frac{100}{1 + \text{RS}_t}$$
$$\text{RS}_t = \frac{\text{Average Gain}_t}{\text{Average Loss}_t}$$

**交易信号**：

买入信号：
$$\text{BuySignal} = (P_t \leq \text{Lower}_t) \land (\text{RSI}_t < 30)$$

卖出信号：
$$\text{SellSignal} = (P_t \geq \text{Upper}_t) \land (\text{RSI}_t > 70)$$

**布林带带宽（BandWidth）**：
$$\text{BW}_t = \frac{\text{Upper}_t - \text{Lower}_t}{\text{Middle}_t}$$

带宽收窄预示波动率即将放大。
### Python代码示例
```python
import numpy as np
import pandas as pd
from typing import Dict, List, Tuple

class BollingerBandReversionStrategy:
    """布林带均值回归策略"""
    
    def __init__(self, bb_period: int = 20,
                 bb_std: float = 2.0,
                 rsi_period: int = 14,
                 rsi_oversold: float = 30,
                 rsi_overbought: float = 70,
                 use_bandwidth_confirm: bool = True):
        self.bb_period = bb_period
        self.bb_std = bb_std
        self.rsi_period = rsi_period
        self.rsi_oversold = rsi_oversold
        self.rsi_overbought = rsi_overbought
        self.use_bandwidth_confirm = use_bandwidth_confirm
        
    def calculate_bollinger_bands(self, prices: pd.Series) -> Tuple[pd.Series, pd.Series, pd.Series]:
        """计算布林带"""
        middle = prices.rolling(window=self.bb_period).mean()
        std = prices.rolling(window=self.bb_period).std()
        
        upper = middle + self.bb_std * std
        lower = middle - self.bb_std * std
        
        return upper, middle, lower
    
    def calculate_rsi(self, prices: pd.Series) -> pd.Series:
        """计算RSI指标"""
        delta = prices.diff()
        
        gain = delta.where(delta > 0, 0)
        loss = (-delta).where(delta < 0, 0)
        
        avg_gain = gain.rolling(window=self.rsi_period).mean()
        avg_loss = loss.rolling(window=self.rsi_period).mean()
        
        rs = avg_gain / (avg_loss + 1e-10)
        rsi = 100 - (100 / (1 + rs))
        
        return rsi
    
    def calculate_bandwidth(self, prices: pd.Series) -> pd.Series:
        """计算布林带带宽"""
        upper, middle, lower = self.calculate_bollinger_bands(prices)
        bandwidth = (upper - lower) / (middle + 1e-10)
        return bandwidth
    
    def calculate_percent_b(self(self, prices: pd.Series) -> pd.Series:
        """计算%B指标"""
        upper, middle, lower = self.calculate_bollinger_bands(prices)
        percent_b = (prices - lower) / (upper - lower + 1e-10)
        return percent_b
    
    def calculate_atr(self, prices: pd.Series, period: int = 14) -> pd.Series:
        """计算ATR用于止损"""
        high = prices.rolling(window=period).max()
        low = prices.rolling(window=period).min()
        
        tr = np.maximum(
            high - low,
            np.maximum(
                abs(high - prices.shift(1)),
                abs(low - prices.shift(1))
            )
        )
        
        atr = tr.rolling(window=period).mean()
        return atr
    
    def generate_signal(self, prices: pd.Series) -> Dict:
        """生成交易信号"""
        upper, middle, lower = self.calculate_bollinger_bands(prices)
        rsi = self.calculate_rsi(prices)
        bandwidth = self.calculate_bandwidth(prices)
        percent_b = self.calculate_percent_b(prices)
        
        current_price = prices.iloc[-1]
        current_rsi = rsi.iloc[-1]
        current_upper = upper.iloc[-1]
        current_lower = lower.iloc[-1]
        current_middle = middle.iloc[-1]
        current_bw = bandwidth.iloc[-1]
        current_pb = percent_b.iloc[-1]
        
        signal = {
            'price': current_price,
            'position': 'middle',
            'rsi': current_rsi,
            'percent_b': current_pb,
            'bandwidth': current_bw
        }
        
        bw_ma = bandwidth.rolling(20).mean().iloc[-1]
        bw_squeeze = current_bw < bw_ma * 0.8
        
        if current_price <= current_lower:
            if current_rsi < self.rsi_oversold:
                if self.use_bandwidth_confirm and not bw_squeeze:
                    signal['position'] = 'long'
                    signal['confidence'] = 0.5
                else:
                    signal['position'] = 'long'
                    signal['confidence'] = 0.8
            else:
                signal['position'] = 'watch_long'
        elif current_price >= current_upper:
            if current_rsi > self.rsi_overbought:
                if self.use_bandwidth_confirm and not bw_squeeze:
                    signal['position'] = 'short'
                    signal['confidence'] = 0.5
                else:
                    signal['position'] = 'short'
                    signal['confidence'] = 0.8
            else:
                signal['position'] = 'watch_short'
        elif current_price >= current_middle * 1.02:
            signal['position'] = 'near_upper'
        elif current_price <= current_middle * 0.98:
            signal['position'] = 'near_lower'
            
        return signal
    
    def calculate_stop_loss(self, entry_price: float,
                           direction: str,
                           atr: float) -> Tuple[float, float]:
        """计算止损位"""
        if direction == 'long':
            stop_loss = entry_price - 2 * atr
            take_profit = entry_price + 3 * atr
        else:
            stop_loss = entry_price + 2 * atr
            take_profit = entry_price - 3 * atr
            
        return stop_loss, take_profit
    
    def calculate_position_size(self, portfolio_value: float,
                               entry_price: float,
                               atr: float,
                               max_loss_pct: float = 0.02) -> float:
        """仓位管理"""
        risk_amount = portfolio_value * max_loss_pct
        
        share_size = risk_amount / (2 * atr + 1e-10)
        
        position_value = share_size * entry_price
        
        return min(position_value, portfolio_value * 0.1)
    
    def run_backtest(self, price_data: pd.DataFrame,
                   initial_capital: float = 1000000) -> Dict:
        """运行回测"""
        prices = price_data['close']
        atr = self.calculate_atr(prices)
        
        portfolio_values = [initial_capital]
        signals = []
        position = 0
        entry_price = 0
        entry_date = None
        
        for i in range(self.bb_period + 1, len(prices)):
            window_prices = prices.iloc[:i]
            signal = self.generate_signal(window_prices)
            
            current_price = prices.iloc[i]
            current_atr = atr.iloc[i]
            
            if position == 0:
                if signal['position'] == 'long':
                    position = 1
                    entry_price = current_price
                    entry_date = prices.index[i]
                elif signal['position'] == 'short':
                    position = -1
                    entry_price = current_price
                    entry_date = prices.index[i]
            else:
                stop_loss, take_profit = self.calculate_stop_loss(
                    entry_price,
                    'long' if position == 1 else 'short',
                    current_atr
                )
                
                should_exit = False
                
                if position == 1:
                    if current_price <= stop_loss:
                        should_exit = True
                    elif current_price >= take_profit:
                        should_exit = True
                else:
                    if current_price >= stop_loss:
                        should_exit = True
                    elif current_price <= take_profit:
                        should_exit = True
                        
                holding_days = (prices.index[i] - entry_date).days if entry_date else 0
                if holding_days > 30:
                    should_exit = True
                    
                if should_exit:
                    ret = position * (current_price - entry_price) / entry_price
                    new_value = portfolio_values[-1] * (1 + ret)
                    portfolio_values.append(new_value)
                    position = 0
                else:
                    portfolio_values.append(portfolio_values[-1])
                    
            signals.append(signal)
            
        return {
            'portfolio_values': portfolio_values,
            'signals': signals,
            'total_return': (portfolio_values[-1] - initial_capital) / initial_capital,
            'final_position': position
        }
```
### 数据来源
- Reddit: r/quant, r/algorithmictrading, r/TechnicalAnalysis
- 小红书: 技术分析、量化投资、均线策略
- 知乎: 量化交易、技术分析、均值回归
- 其他: Investopedia, TradingView, MetaTrader
